cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java

```diff
@@ -107,7 +115,13 @@ protected void runOneIteration() throws Exception {
     if (threadPoolExecutor != null) {
       for (EventReader<StartProgramEvent> reader : readers) {
         threadPoolExecutor.execute(() -> {
-          processEvents(reader);
+          // There will be race conditions
+          RunRecordMonitorService.Counter counter = runRecordMonitorService.getImpreciseCount();
+          // Only attempt to process event if there is no max or the current count is less than max
+          if (maxConcurrentRuns < 0 ||
+                  counter.getLaunchingCount() + counter.getRunningCount() < maxConcurrentRuns) {
```

Comment:
I suggest we introduce some configurable value such that we have some wiggle room when in fact perform a pull. For instance, we don't precisely check launching + running < maxConcurrentRuns. Instead, we check launching + running < 90% of maxConcurrentRuns . This way, we have 10% wiggle room from this check until we do actual pull to avoid 429. 

The wiggle room should also consider number of events we'll pull. E.g., if we pull 10 events at every loop, then we should make sure we have room for 10 starts. Right?